From d6ff2b336bf1633f56aab781f40f3e79b408a8c8 Mon Sep 17 00:00:00 2001
From: Matthias Schiffer <mschiffer@universe-factory.net>
Date: Fri, 1 Apr 2016 23:16:13 +0200
Subject: [PATCH 21/93] ar71xx: check both HWID and HWREV on upgrades of
 TP-LINK devices

There's no reason for us to be more lenient than the stock firmware, so
better check the HWREV as well to avoid bricked devices.

Signed-off-by: Matthias Schiffer <mschiffer@universe-factory.net>

Backport of r49105
---
 target/linux/ar71xx/base-files/lib/upgrade/platform.sh | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/target/linux/ar71xx/base-files/lib/upgrade/platform.sh b/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
index 762cd760ce..024e493094 100755
--- a/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
+++ b/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
@@ -66,6 +66,10 @@ tplink_get_image_hwid() {
 	get_image "$@" | dd bs=4 count=1 skip=16 2>/dev/null | hexdump -v -n 4 -e '1/1 "%02x"'
 }
 
+tplink_get_image_mid() {
+	get_image "$@" | dd bs=4 count=1 skip=17 2>/dev/null | hexdump -v -n 4 -e '1/1 "%02x"'
+}
+
 tplink_get_image_boot_size() {
 	get_image "$@" | dd bs=4 count=1 skip=37 2>/dev/null | hexdump -v -n 4 -e '1/1 "%02x"'
 }
@@ -373,13 +377,17 @@ platform_check_image() {
 		}
 
 		local hwid
-		local imageid
+		local mid
+		local imagehwid
+		local imagemid
 
 		hwid=$(tplink_get_hwid)
-		imageid=$(tplink_get_image_hwid "$1")
+		mid=$(tplink_get_mid)
+		imagehwid=$(tplink_get_image_hwid "$1")
+		imagemid=$(tplink_get_image_mid "$1")
 
-		[ "$hwid" != "$imageid" ] && {
-			echo "Invalid image, hardware ID mismatch, hw:$hwid image:$imageid."
+		[ "$hwid" != "$imagehwid" -o "$mid" != "$imagemid" ] && {
+			echo "Invalid image, hardware ID mismatch, hw:$hwid $mid image:$imagehwid $imagemid."
 			return 1
 		}
 
-- 
2.11.0

